muxers/yamux/src/lib.rs

```diff
@@ -252,62 +283,69 @@ impl WindowUpdateMode {
     /// > receive buffer being larger than the window size.
     pub fn on_read() -> Self {
-        WindowUpdateMode(yamux::WindowUpdateMode::OnRead)
+        WindowUpdateMode(yamux012::WindowUpdateMode::OnRead)
     }
 }
 
 impl Config {
     /// Creates a new `YamuxConfig` in client mode, regardless of whether
     /// it will be used for an inbound or outbound upgrade.
+    #[deprecated(note = "Will be removed with the next breaking release.")]
     pub fn client() -> Self {
-        Self {
-            mode: Some(yamux::Mode::Client),
+        Self(Either::Left(Config012 {
+            mode: Some(yamux012::Mode::Client),
             ..Default::default()
-        }
+        }))
     }
 
     /// Creates a new `YamuxConfig` in server mode, regardless of whether
     /// it will be used for an inbound or outbound upgrade.
+    #[deprecated(note = "Will be removed with the next breaking release.")]
     pub fn server() -> Self {
-        Self {
-            mode: Some(yamux::Mode::Server),
+        Self(Either::Left(Config012 {
+            mode: Some(yamux012::Mode::Server),
             ..Default::default()
-        }
+        }))
     }
 
-    /// Sets the size (in bytes) of the receive window per substream.
-    pub fn set_receive_window_size(&mut self, num_bytes: u32) -> &mut Self {
-        self.inner.set_receive_window(num_bytes);
+    fn set(&mut self, f: impl FnOnce(&mut yamux012::Config) -> &mut yamux012::Config) -> &mut Self {
+        let cfg012 = match self.0.as_mut() {
+            Either::Left(c) => &mut c.inner,
+            Either::Right(_) => {
+                self.0 = Either::Left(Config012::default());
+                &mut self.0.as_mut().unwrap_left().inner
+            }
+        };
+
+        f(cfg012);
+
         self
     }
 
+    /// Sets the size (in bytes) of the receive w``indow per substream.
+    #[deprecated(
+        note = "Will be replaced in the next breaking release with a connection receive window size limit."
+    )]
+    pub fn set_receive_window_size(&mut self, num_bytes: u32) -> &mut Self {
+        self.set(|cfg| cfg.set_receive_window(num_bytes))
+    }
+
     /// Sets the maximum size (in bytes) of the receive buffer per substream.
+    #[deprecated(note = "Will be removed with the next breaking release.")]
     pub fn set_max_buffer_size(&mut self, num_bytes: usize) -> &mut Self {
-        self.inner.set_max_buffer_size(num_bytes);
-        self
+        self.set(|cfg| cfg.set_max_buffer_size(num_bytes))
     }
 
     /// Sets the maximum number of concurrent substreams.
     pub fn set_max_num_streams(&mut self, num_streams: usize) -> &mut Self {
-        self.inner.set_max_num_streams(num_streams);
-        self
+        self.set(|cfg| cfg.set_max_num_streams(num_streams))
     }
 
     /// Sets the window update mode that determines when the remote
     /// is given new credit for sending more data.
+    #[deprecated(note = "Will be removed with the next breaking release.")]
```

Comment:
Maybe this should explain that `WindowUpdateMode::on_read` breaks backpressure and will thus be removed.